lib/queue/job.js

```diff
@@ -788,29 +805,66 @@ Job.prototype.save = function( fn ) {
   // update
   if( this.id ) return this.update(fn);
 
+  // unique id?
+  if( this._unique ) return this.saveUnique(fn);
+
   // incr id
   client.incr(client.getKey('ids'), function( err, id ) {
     if( err ) return fn(err);
-    // add the job for event mapping
-    self.id = id;
-    self.zid = client.createFIFO(id);
-    self.subscribe(function() {
-      self._state     = self._state || (this._delay ? 'delayed' : 'inactive');
-      if( max ) { self.set('max_attempts', max); }
-      client.sadd(client.getKey('job:types'), self.type, noop);
-      self.set('type', self.type);
-      var now         = Date.now();
-      self.created_at = now;
-      self.set('created_at', self.created_at);
-      self.promote_at = now + (self._delay || 0);
-      self.set('promote_at', self.promote_at);
-      self.update(fn);
-    }.bind(this));
+    self.saveWithId(id, fn);
   }.bind(this));
   return this;
 };
 
 /**
+ * Save the job, optionally invoking the callback `fn(err)`.
+ *
+ * @param {number} id
+ * @param {Function} fn
+ * @return {Job} for chaining
+ * @api public
+ */
+
+Job.prototype.saveUnique = function( id, fn ) {
+  throw new Error('TODO: Check/set a hash, mapping unique identifiers => integer id');
```

Comment:
After a bit more thought, it seems like I should just do `HSET` right away, and use the return value to know if it was present or not.  This should solve the race condition.

> Return value
Integer reply, specifically:
1 if field is a new field in the hash and value was set.
**0 if field already exists in the hash and the value was updated.**

To support this though, I'd need to unconditionally call `client.incr`, but I assume it doesn't matter if the incremented value doesn't actually get used